### PR TITLE
Added positive tests in stake pool and fixed several bugs in instructions

### PR DESCRIPTION
**Pull request by ysavchenko**
Positive tests for:
- withdraw
- claim
- set staking authority
- set owner

Fixed bugs when forming claim and set owner instructions.

### CODE REVIEW

File: stake-pool/program/src/processor.rs [hunk 1]
```diff
@@ -941,70 +1047,241 @@ mod tests {
...
-        let mut stake_account_account = Account::new(stake_balance, 100, &stake_program_id());
-        // TODO: Set stake account Withdrawer authority to pool_info.deposit_authority_key
+        let deposit_info = do_deposit(&mut pool_info, stake_balance);
 
-        // Create account to receive minted tokens
-        let (token_receiver_key, mut token_receiver_account) = create_token_account(
+        // Test stake pool balance
+        let state = State::deserialize(&pool_info.pool_account.data).unwrap();
+        match state {
+            State::Unallocated => panic!("Stake pool state is not initialized after deposit"),
```
Comment (joncinque) [inline stake-pool/program/src/processor.rs:1055]: nit: Prefer to use the `matches!` macro instead https://doc.rust-lang.org/std/macro.matches.html

File: stake-pool/program/src/processor.rs [hunk 2]
```diff
@@ -941,70 +1047,241 @@ mod tests {
...
+                .expect("User token account is not initialized after deposit");
+        assert_eq!(user_token_state.amount, user_token_balance);
+        let fee_token_state = SplAccount::unpack_from_slice(&pool_info.owner_fee_account.data)
+            .expect("Fee token account is not initialized after deposit");
+        assert_eq!(fee_token_state.amount, fee_token_balance);
+
+        // Test mint total issued tokens
+        let mint_state = SplMint::unpack_from_slice(&pool_info.mint_account.data)
+            .expect("Mint account is not initialized after deposit");
+        assert_eq!(mint_state.supply, stake_balance);
```
Comment (joncinque) [inline stake-pool/program/src/processor.rs:1074]: It may be useful to add a TODO to check the results of the stake instruction too

File: stake-pool/program/src/processor.rs [hunk 3]
```diff
@@ -941,70 +1047,241 @@ mod tests {
...
                 &mut Account::default(),
                 &mut Account::default(),
                 &mut Account::default(),
             ],
         )
-        .expect("Error on stake pool deposit");
+        .expect("Error on withdraw");
 
-        // Test stake pool balance
+        let fee_amount = stake_balance * FEE_DEFAULT.numerator / FEE_DEFAULT.denominator;
```
Comment (joncinque) [inline stake-pool/program/src/processor.rs:1129]: You can use the `fee` on `pool_info` to calculate this instead

File: stake-pool/program/src/processor.rs [hunk 4]
```diff
@@ -941,70 +1047,241 @@ mod tests {
...
+        assert_eq!(
+            user_token_state.amount,
+            stake_balance - fee_amount - withdraw_amount
+        );
+
+        // Check stake pool token amounts
         let state = State::deserialize(&pool_info.pool_account.data).unwrap();
         match state {
-            State::Unallocated => panic!("Stake pool state is not initialized after deposit"),
+            State::Unallocated => panic!("Stake pool state is not initialized after withdraw"),
```
Comment (joncinque) [inline stake-pool/program/src/processor.rs:1142]: You can do `assert!(matches!(...` here too

File: stake-pool/program/src/processor.rs [hunk 5]
```diff
@@ -941,70 +1047,241 @@ mod tests {
...
+    #[test]
+    fn test_withdraw() {
+        let mut pool_info = create_stake_pool_default();
+
+        let user_withdrawer_key = Pubkey::new_unique();
+
+        let stake_balance = sol_to_lamports(20.0);
+        let mut deposit_info = do_deposit(&mut pool_info, stake_balance);
+
+        let withdraw_amount = sol_to_lamports(5.0);
```
Comment (joncinque) [inline stake-pool/program/src/processor.rs:1085]: nit: since `withdraw_amount` is in pool tokens and not lamports, could you hard-code a number instead?

File: stake-pool/program/src/processor.rs [hunk 6]
```diff
@@ -941,70 +1047,241 @@ mod tests {
...
 
-        // Test mint total issued tokens
-        let mint_state = SplMint::unpack_from_slice(&pool_info.mint_account.data)
-            .expect("Mint account is not initialized after deposit");
-        assert_eq!(mint_state.supply, stake_balance);
+        let stake_balance = sol_to_lamports(20.0);
+        let mut deposit_info = do_deposit(&mut pool_info, stake_balance);
+
+        // Need to deposit more to cover deposit fee
+        let fee_amount = stake_balance * FEE_DEFAULT.numerator / FEE_DEFAULT.denominator;
```
Comment (joncinque) [inline stake-pool/program/src/processor.rs:1159]: nit: You can use the fee on `pool_info` here as well

File: stake-pool/program/src/processor.rs [hunk 7]
```diff
@@ -941,70 +1047,241 @@ mod tests {
...
+                &mut Account::default(),
+                &mut Account::default(),
+            ],
+        )
+        .expect("Error on claim");
+
+        let user_token_state =
+            SplAccount::unpack_from_slice(&extra_deposit_info.token_receiver_account.data)
+                .expect("User token account is not initialized after withdraw");
+        assert_eq!(user_token_state.amount, 0);
```
Comment (joncinque) [inline stake-pool/program/src/processor.rs:1211]: nit: you can add a TODO here as well to check the stake account ownership

File: stake-pool/program/src/processor.rs [hunk 8]
```diff
@@ -941,70 +1047,241 @@ mod tests {
...
+                &mut pool_info.owner_account,
+                &mut new_owner_account,
+                &mut new_owner_fee.account,
+            ],
+        )
+        .expect("Error on set_owner");
+
+        let state = State::deserialize(&pool_info.pool_account.data).unwrap();
+        match state {
+            State::Unallocated => panic!("Stake pool state is not initialized after set owner"),
```
Comment (joncinque) [inline stake-pool/program/src/processor.rs:1280]: nit: `assert!(matches!(...` here too

**Review by joncinque — COMMENTED** [review]
Looks great, mainly just a few nits

File: stake-pool/program/src/processor.rs [hunk 9]
```diff
@@ -892,20 +943,75 @@ mod tests {
...
+            ],
+        )
+        .expect("Error on stake pool deposit");
+
+        DepositInfo {
+            stake_account_key,
+            stake_account_account,
+            token_receiver_key: token.key,
+            token_receiver_account: token.account.clone(),
+            token_owner_key: token.owner,
```
[the remark this answers is not in the document]
Comment (ysavchenko) [reply stake-pool/program/src/processor.rs:1008]: Actually there are two deposit functions: one accepts token account as a parameter and another creates new token account and passes it to the first function (`do_deposit` and `do_deposit_custom`). We can make `do_deposit_custom` return only stake account info (key/account) but it is better to have them both return the same DepositInfo struct (to make code calling these two functions similar).

[the remark this answers is not in the document]
Comment (joncinque) [reply stake-pool/program/src/processor.rs:1011]: I see what you mean.  To be honest, it was a bit hard to follow some of the tests because of the account creation in `do_deposit`, so I would explicitly create the account outside, and only have `do_deposit_custom`.  What do you think?

[the remark this answers is not in the document]
Comment (ysavchenko) [reply stake-pool/program/src/processor.rs:1011]: Actually it was the other way round :) We did do_deposit, which is used in deposit/withdraw/claim tests. And then realized that we need to collect tokens for two deposits to the same accounts, so we've added do_deposit_custom to allow that. And I think in remaining tests we will mostly use do_deposit which will create token account by itself.

But you of course have the last word, if you think it will be clearer if token receiver was created outside of the deposit we will do it this way :)

[the remark this answers is not in the document]
Comment (joncinque) [reply stake-pool/program/src/processor.rs:1011]: Creating an account in `do_deposit` still seems confusing to me, so I'd prefer that we only have the version that takes `&mut TokenInfo`, and then also removing the `token_*` fields from `DepositInfo`.  That way, you don't need to worry about data duplication.

**Comment by joncinque** [comment]
Looks great, thanks for the refinements!

**Review by joncinque — APPROVED** [review]
